test/pop_manager_test.erl

```diff
@@ -115,37 +121,80 @@ extract_new_block_messages() ->
 	    []
     end.
 
+extract_send_full_blocks() ->
+    receive
+	{net, send_full_blocks, {old, [Block]}} -> 
+	    [Block | extract_send_full_blocks()];
+	_Any ->
+	    fail
+    after 0 -> 
+	    []
+    end.
+    
+
 no_more_messages() ->
     receive
 	_Any ->
 	    false
     after 0 -> 
 	    true
     end.
+
+make_transaction(PrivateKey, PublicKey, Nonce, ChainId) ->
+    T0 = #{
+	   game_data => undefined,
+	   nonce => Nonce,
+	   player_id => PublicKey,
+	   consensus_data => #{
+			       signature => undefined,
+			       chain_id => ChainId
+			      }
+	  },
+    Hash = my_crypto:hash(my_serializer:serialize_object(T0)),
+    Signature = my_crypto:sign(Hash, PrivateKey),
+    CD = maps:get(consensus_data, T0),
+    T1 = T0#{
+	     consensus_data := CD#{signature := Signature}
+	    },
+
+    T1.
     
 
 fancy_test() ->
-    PM = make_pm(),
+    {PM, PrivateKey, PublicKey} = make_pm(),
     PC = PM#pop_manager.pop_chain,
     Time = 2000,
 
     A0 = pop_chain:get_head_block(PC),
 
-    generate_chain(PC, 1, 0),
+    ?assertEqual(maps:get(this_id, A0), genesis),
 
-    {PC_A, [A1, A2]} = generate_chain(PC, 2, 0),
+    generate_chain(PC, 1, 0, PrivateKey),
 
-    {PC_B, [B1, B2, B3, B4]} = generate_chain(PC_A, 4, 0),
+    {PC_A, [A1, A2]} = generate_chain(PC, 2, 0, PrivateKey),
+
+    {PC_B, [B1, B2, B3, B4]} = generate_chain(PC_A, 4, 0, PrivateKey),
+
+    {PC_C, [C1, C2, C3, C4, C5]} = generate_chain(PC_A, 5, 1, PrivateKey),
+
+    %% We have constructed test chain:
+    %% 
+    %%                B1 - B2 - B3 - B4
+    %%              /
+    %% A0 - A1 - A2 
+    %%              \
+    %%                C1 - C2 - C3 - C4 - C5
+    %% 
```

Comment:
except for this part, this part is awesome

Reply:
good except needs more fish and salad emojis  🥗 🐟

Reply:
OMG EACH HASH NEEDS TO BE AN EMOJI

gonna rewrite my hash_debug function

Reply:
https://gist.github.com/rxaviers/7360908

:money_with_wings: :crystal_ball: :waning_crescent_moon: :water_buffalo: :stuck_out_tongue_closed_eyes:

Reply:
https://api.github.com/emojis

Reply:
watch out for emoji collisions :o!

Reply:
https://www.youtube.com/watch?v=u011XHmYq1Q

Reply:
SHA256 hash is equivalent to approx 11 emojis